container/go/pkg/compat/overrideConfig.go

```diff
@@ -0,0 +1,545 @@
+			return fmt.Sprintf("%v", errors.Wrap(err, "failed to create new error"))
+		}
+		return fmt.Sprintf("%v", err)
+	}
+
+	expandedVar := os.Expand(value, mapper)
+	if errorMet {
+		return "", errors.New(fmt.Sprintf("variable substitution of %s was unsuccessful", value))
+	}
+	return expandedVar, nil
+}
+
+// getCreationTime returns the correct creation time for which to override the current config file.
+func getCreationTime(overrideInfo *OverrideConfigOpts) (time.Time, error) {
+	// createTime stores the input creation time with macros substituted.
+	var createTime string
+	var creationTime time.Time
+	var err error
+	// Parse for specific time formats.
+	if overrideInfo.CreationTimeString != "" {
+		var unixTime float64
+		// Use stamp to as preliminary replacement.
+		if createTime, err = Stamp(overrideInfo.CreationTimeString, overrideInfo.StampInfoFile); err != nil {
+			return time.Unix(0, 0), errors.Wrapf(err, "unable to format creation time from BUILD_TIMESTAMP macros")
+		}
+		// If creationTime is parsable as a floating point type, assume unix epoch timestamp.
+		// otherwise, assume RFC 3339 date/time format.
+		unixTime, err := strconv.ParseFloat(createTime, 64)
+		// Assume RFC 3339 date/time format. No err means it is parsable as floating point.
+		if err == nil {
+			// Ensure that the parsed time is within the floating point range.
+			// Values > 1e11 are assumed to be unix epoch milliseconds.
+			if unixTime > 1.0e+11 {
+				unixTime = unixTime / 1000.0
+			}
+			// Construct a RFC 3339 date/time from Unix epoch.
+			sec, dec := math.Modf(unixTime)
+			creationTime = time.Unix(int64(sec), int64(dec*(1e9))).UTC()
+		} else {
+			creationTime, err = time.Parse(time.RFC3339, createTime)
+			if err != nil {
+				return time.Unix(0, 0), errors.Wrapf(err, "failed to parse the %s into float, so assuming RFC3339", createTime)
+			}
+		}
+	}
+	return creationTime, nil
+}
+
+// updateConfigInfo returns an updated version of the input infoToStamp.
+func updateConfigInfo(overrideInfo *OverrideConfigOpts, infoToStamp []string) ([]string, error) {
+	var infoStamped = infoToStamp
+	for i, entry := range infoToStamp {
+		stampedEntry, err := Stamp(entry, overrideInfo.StampInfoFile)
+		if err != nil {
+			return []string{}, errors.Wrapf(err, "unable to perform substitutions to Env variable %s", entry)
+		}
+		infoStamped[i] = stampedEntry
+	}
+
+	return infoStamped, nil
+}
+
+// resolveEnvironment returns a string array with fully substituted environment variables.
+func resolveEnvironment(overrideInfo *OverrideConfigOpts, environMap map[string]string) ([]string, error) {
+	var baseEnvMap = make(map[string]string)
+	var err error
+	if len(overrideInfo.ConfigFile.Config.Env) > 0 {
+		baseEnvMap, err = keyValueToMap(overrideInfo.ConfigFile.Config.Env)
+		if err != nil {
+			return []string{}, errors.Wrapf(err, "error converting Config env array %v to map", overrideInfo.ConfigFile.Config.Env)
+		}
+	}
+
+	for k, v := range environMap {
```

Comment:
This loop is modifying the `baseEnvMap` as it's looping through it. I think this is problematic because iteration through Go maps does not have any pre-defined order. So imagine a baseEnvMap with
```
"PATH": "$PATH:a",
"FOO": "$PATH:b",
```

If the map iterates over "PATH" followed by "FOO" we get:
```
"PATH": ":a",
"FOO": ":a:b",
```

However, if the map iterates over "FOO" followed by "PATH" we get:
```
"PATH": ":a",
"FOO": ":b",
```

The iteration order can change arbitrarily depending on the contents of the map. I think the correct implementation would not modify the baseEnvMap while resolving the variables. It should create a separate output map that contains the resolved values. So something like:
```
result := make(map[string]string)
for k, v := range environMap {
  var expanded string
  if expanded, err = resolveVariables(v, baseEnvMap); err != nil {
    return nil, errors.Wrapf(err, "unable to resolve environment variables in %s with content mapping at %s", k, v)
  }
  result[k] = expanded
```